File: python/队列/406. 根据身高重建队列.py
from typing import List
# 解题思路：先排序再插入
#      * 1.排序规则：按照先H高度降序，K个数升序排序
#      * 2.遍历排序后的数组，根据K插入到K的位置上
#      *
#      * 核心思想：高个子先站好位，矮个子插入到K位置上，前面肯定有K个高个子，矮个子再插到前面也满足K的要求

#  [7,0], [7,1], [6,1], [5,0], [5,2], [4,4]
#  再一个一个插入。
#  [7,0]
#  [7,0], [7,1]
#  [7,0], [6,1], [7,1]
#  [5,0], [7,0], [6,1], [7,1]
#  [5,0], [7,0], [5,2], [6,1], [7,1]
#  [5,0], [7,0], [5,2], [6,1], [4,4], [7,1]

class Solution:
    def reconstructQueue(self, people: List[List[int]]) -> List[List[int]]:
        people.sort(key=lambda x: (-x[0], x[1]))
        res = []
        for i in people:
            res.insert(i[1], i)
        return res

s = Solution()
people = [[7,0], [4,4], [7,1], [5,0], [6,1], [5,2]]
print(s.reconstructQueue(people))

File: go/queue/406. 根据身高重建队列.go
package queue

import "sort"

func reconstructQueue(people [][]int) [][]int {
	// 排序，先按照身高h排序，如果身高一样高，那就按照人数k排序
	sort.Slice(people, func(i, j int) bool {
		a, b := people[i], people[j]
		return a[0] > b[0] || a[0] == b[0] && a[1] < b[1]
	})

	// 按照k值插入到index=k的地方，index之后的往后移动
	for i, p := range people {
		copy(people[p[1]+1:i+1], people[p[1]:i+1])
		people[p[1]] = p
	}

	return people
}
